### PR TITLE
kv: remove btree, and use rbtree to implement membuffer.

### PR DESCRIPTION
**Pull request by coocood**
LLRB tree has better performance than 'memkv' btree, and similar performance than 'leveldb/memdb'.
But it has much lower creation cost (108 ns vs 14235 ns), so lazyMembuffer can be removed.

```
BenchmarkRBTreeBufferSequential-4	      10	 195011273 ns/op	 8480078 B/op	  210007 allocs/op
BenchmarkRBTreeBufferRandom-4    	       3	 397533394 ns/op	 9600144 B/op	  233348 allocs/op
BenchmarkMemDbBufferSequential-4 	      10	 149774580 ns/op	13374417 B/op	      14 allocs/op
BenchmarkMemDbBufferRandom-4     	       3	 345256714 ns/op	19694016 B/op	      37 allocs/op
BenchmarkBTreeBufferSequential-4	       2	 508861362 ns/op	16925848 B/op	  600409 allocs/op
BenchmarkBTreeBufferRandom-4    	       2	 880584695 ns/op	17475352 B/op	  600496 allocs/op

BenchmarkRBTreeIter-4            	      20	  94366649 ns/op	 4800105 B/op	  200004 allocs/op
BenchmarkBTreeIter-4            	     200	   7570513 ns/op	     144 B/op	       2 allocs/op
BenchmarkMemDbIter-4             	     100	  12311667 ns/op	     192 B/op	       3 allocs/op

BenchmarkRBTreeCreation-4        	20000000	       108 ns/op	      40 B/op	       3 allocs/op
BenchmarkBTreeCreation-4        	10000000	       130 ns/op	      72 B/op	       2 allocs/op
BenchmarkMemDbCreation-4        	  100000	     14235 ns/op	    9848 B/op	       6 allocs/op
```

For simplicity, `rbTreeIter` is not buffered, so the performance is not as good, we can optimize this when we need to later.

### CODE REVIEW

**Comment by ngaut** [comment]
Please add some SQL bench results.

**Comment by ngaut** [comment]
@coocood Does #1049 related to this PR ?

**Comment by coocood** [comment]
@ngaut This PR doesn't improve the INSERT performance because MemDB uses a pool, creation time doesn't matter.

**Comment by coocood** [comment]
Tested several round, for InsertWithIndex, ns/op is around 220,000, the difference is negligible, but for `InsertWithNoIndex`, MemDB is about 10% faster, 170 us than RBTree's 190 us.

So, I kept the MemDB as default MemBuffer, only deleted btreeBuffer.

File: kv/rbtree_buffer.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	if len(v) == 0 {
+		return errors.Trace(ErrCannotSetNilValue)
+	}
+	m.tree.ReplaceOrInsert(&pairItem{key: k, value: v})
+	return nil
+}
+
+// Delete removes the entry from buffer with provided key.
+func (m *rbTreeBuffer) Delete(k Key) error {
+	m.tree.ReplaceOrInsert(&pairItem{key: k, value: nil})
```
Comment (shenli) [inline kv/rbtree_buffer.go]: Why not use m.tree.Delete()?

Comment (coocood) [inline kv/rbtree_buffer.go]: Because MVCC

**Comment by shenli** [comment]
LGTM

File: kv/rbtree_buffer.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+// NewRBTreeBuffer creates a new rbTreeBuffer.
+func NewRBTreeBuffer() MemBuffer {
+	return &rbTreeBuffer{tree: llrb.New()}
+}
+
+// Seek creates an Iterator.
+func (m *rbTreeBuffer) Seek(k Key) (Iterator, error) {
+	it := &rbTreeIter{tree: m.tree, seek: k}
+	it.Next()
```
Comment (zimulala) [inline kv/rbtree_buffer.go]: Need check the error ?

Comment (coocood) [inline kv/rbtree_buffer.go]: `rbTreeIter.Next` never returns error.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**